1.segment_images/scripts/3.segmentation_decoupling.py

```diff
@@ -185,76 +132,75 @@ def decouple_masks(self):
+    merged_df = get_combinations_of_indices(df, distance_threshold=distance_threshold)
+    # combine dfs for each window index
+    for window_index in range(window_size + 1):
+        merged_df = merge_sets(merged_df)
+    if not merged_df.empty:
+        merged_df.loc[:, "slice"] = zslice
+        reassembled_masks = reassemble_each_mask(
+            merged_df, original_img_shape=original_imgs.shape
+        )
+        masks_dict[zslice] = reassembled_masks
+    else:
+        print(f"Warning: merged_df is empty for zslice {zslice}")
 
-
-def call_mask_decoupling(z_stack_index, z_stack_mask):
-    decouple = DecoupleSlidingWindowMasks(
-        lambda_IOU_threshold=0.8, image_stack=z_stack_mask
+    reassembled_masks = reassemble_each_mask(
+        merged_df, original_img_shape=original_imgs.shape
     )
-    new_image = decouple.decouple_masks()
-    return z_stack_index, new_image
-
-
-# process each z slice in parallel
-with Pool(num_cores) as p:
-    results = p.starmap(call_mask_decoupling, reconstruction_dict.items())
+    masks_dict[zslice] = reassembled_masks
 
-# reconstruct the masks into a single image (z-stack)
-reconstructed_masks = np.zeros(
-    (original_z_slice_count, original_imgs.shape[1], original_imgs.shape[2])
-)
-for index, new_image in results:
-    reconstructed_masks[index] = new_image
-# cast the reconstructed masks to int8
-reconstructed_masks = reconstructed_masks.astype(np.uint8)
 
+# In[6]:
 
-# In[7]:
 
+# convert the masks_dict to a numpy array
+masks = np.zeros(
+    (original_z_slice_count, original_img_shape[1], original_img_shape[2]),
+    dtype=np.uint16,
+)
+masks = np.array(list(masks_dict.values()))
+if get_number_of_unique_labels(masks) > 255:
+    masks = masks.astype(np.uint16)
+else:
+    masks = masks.astype(np.uint8)
 
-# # save the masks
-print(reconstructed_masks.shape)
-print(reconstructed_masks.dtype)
-print(reconstructed_masks[0])
-print(reconstructed_masks.max())
-print(np.unique(reconstructed_masks))
-# save the masks as tiff
-tifffile.imwrite(mask_file_path, reconstructed_masks)
+# save the masks
+tifffile.imwrite(mask_file_path, masks)
 
 
-# In[8]:
+# In[7]:
 
 
 if in_notebook:
-    # show each z slice of the image and masks
-    for z in range(reconstructed_masks.shape[0]):
-        fig = plt.figure(figsize=(10, 5))
-        plt.subplot(121)
-        plt.imshow(original_imgs[z, :, :], cmap="gray")
-        plt.title("image")
+    for zslice in range(masks.shape[0]):
+        plt.figure(figsize=(10, 10))
+        plt.subplot(1, 2, 1)
+        plt.imshow(original_imgs[zslice], cmap="gray")
+        plt.title(f"original z slice {zslice}")
         plt.axis("off")
-        plt.subplot(122)
-        plt.imshow(reconstructed_masks[z], cmap="gray")
-        plt.title("masks")
+        plt.subplot(1, 2, 2)
+        plt.imshow(masks[zslice], cmap="nipy_spectral")
+        plt.title(f"segmented z slice {zslice}")
         plt.axis("off")
         plt.show()
```

Comment:
Same comment as before, does the color of the masks mean anything and recommend including a code comment on it.